cmd/prune-containers: add a GC script for containers images

jbtrystram opened the pull request:
This script calls skopeo delete to prune image from a remote directory. Currently only supports the FCOS tag structure.

If no duration is specified, no image will be pruned.

See https://github.com/coreos/fedora-coreos-tracker/issues/1367 See https://github.com/coreos/fedora-coreos-pipeline/pull/995

jlebon commented on src/cmd-prune-containers:
I'm surprised by the choice of bash. I think any new command with non-trivial logic should be written in one of the other two languages in this repo (Python or Go).

jlebon replied:
The comments below are only about overall logic and not bash-specific things.

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+
+for tag in $tags; do
+  
+  # ignore the named moving tags ("stable", "next" etc..)
```
I think I would generalize this as "ignore all tags that don't match the build ID format".

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+  # value
+  tag_age=$(( today - $(date -d "${build_date}" +%s) ))
+
+  # test for production streams. Production stream id are 1, 2 and 3
```
I think it'd be nice to follow a model closer to `cosa cloud-prune` where the policy is encoded in a separate YAML file that this command reads in. That would allow us to keep that file in e.g. fedora-coreos-pipeline and possibly down the the line have a similar file for RHCOS.

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+  # test for production streams. Production stream id are 1, 2 and 3
+  if [[ $stream -eq 1 || $stream -eq 2 || $stream -eq 3 ]]; then 
+    if  [[ $tag_age -gt "$prod_age_seconds" ]]; then
+      echo "Production tag ${tag} is older than ${PROD_AGE}, pruning." 
```
This isn't considering barrier releases.

jbtrystram replied:
I guess i would have to pull the `builds.json` for each stream to see if the release is a barrier..
is a LABEL attached to the container to indicate they are barrier releases ?  That would be neat

jlebon replied:
The source of truth for barriers is in https://github.com/coreos/fedora-coreos-streams/tree/main/updates. You can download from there, or from the S3-uploaded version (e.g. https://builds.coreos.fedoraproject.org/updates/testing.json).

jbtrystram replied:
Rewrote in python and now check for barrier releases

jbtrystram commented:
Ah, one last missing bit is : not check barrier release on non-prod streams.

jbtrystram commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,159 @@
+            continue
+
+        # Make sure this is not a barrier release
+        try:
```
Add a condition to skip this barrier release check for non-prod streams

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+    parser.add_argument("--policy", required=True, type=str, help="Path to policy YAML file")
+    parser.add_argument("--dry-run", help="Don't actually delete anything", action='store_true')
+    parser.add_argument("-v", help="Increase verbosity", action='store_true')
+    parser.add_argument("--registry-auth-file", default=os.environ.get("REGISTRY_AUTHFILE"), help="Path to AWS config file")
```
Incorrect help

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+
+def get_update_graph(stream):
+
+    url = f"https://raw.githubusercontent.com/coreos/fedora-coreos-streams/main/updates/{stream}.json"
```
I know I suggested both, but it feels more appropriate to fetch from a service dedicated to serving files than one for code. :)

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+    # Load the policy file
+    with open(args.policy, "r") as f:
+        policy = yaml.safe_load(f)
+    if args.stream not in policy:
+        raise Exception(f"Stream {args.stream} is not defined in policy file.")
+    if 'containers' not in policy[args.stream]:
+        raise Exception(f"No containers section for {args.stream} stream in policy.")
```
I feel like those should just be no-ops instead of hard errors.

jbtrystram replied:
good point

jlebon replied:
Yeah, sorry but no op I mean just `return`.

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+    #    "Tags": [
+    #        "40.20"40.20240301.1.0",.....]}
+    tags_cmd = subprocess.run(["skopeo", "list-tags",
+                              f"docker://{args.repository_url}"],
```
```suggestion
                               f"docker://{args.repository_url}"],
```

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+    # {"Repository": "quay.io/jbtrystramtestimages/fcos",
+    #    "Tags": [
+    #        "40.20"40.20240301.1.0",.....]}
+    tags_cmd = subprocess.run(["skopeo", "list-tags",
+                              f"docker://{args.repository_url}"],
+                              capture_output=True)
+    tags_data = tags_cmd.stdout
+    if tags_cmd.returncode != 0:
+        raise Exception(f"Error retrieving tags list from repo: {tags_cmd.stdout}")
```
```
tags_data = subprocess.check_output(...)
```

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+        update_graph = get_update_graph(args.stream)['releases']
+
+    for tag in tags:
+
```
Let's just here do

```python
if tag in STREAMS:
    continue
...
```

?

I know this would get caught by the `except ValueError` below, but it feels odd to treat tags we _know_ will not be conformant as "unexpected".

That way it makes the "unexpected" log message a better signal we should pay attention to.

jbtrystram replied:
I had done exactly this before in the bash version then you wrote last week : https://github.com/coreos/coreos-assembler/pull/3826#discussion_r1665043870

So.... :) Maybe I can tweak the error message ?

jlebon replied:
Hmm, I don't see them as conflicting. :) The bash version was too permissive, and now this one is too restrictive. Anyway, not a strong opinion!

jbtrystram replied:
Ok I get what you mean now. Now we'll only have "unexpted" logs on tags that are truely not supposed to be there :)

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+
+    stream_id = STREAMS[args.stream]
+    # Get the update graph for stable streams
+    if stream_id in [1, 2, 3]:
```
I think it'd be much clearer to write this using the proper stream names.

```
if args.stream in ['stable', 'testing', 'next']:
```

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+    today = datetime.datetime.now()
+    date_limit = today - relativedelta(days=convert_to_days(policy))
+    if args.v:
+        print(f"This will delete any images older than {date_limit} from the stream {args.stream}")
```
Let's just always print this one.

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+                print(f"Skipping tag {tag} not in {args.stream} stream")
+            continue
+
+        # Make sure this is not a barrier release (for stable streams)
+        if stream_id in [1, 2, 3]:
```
Hmm, that's a duplicate check. Instead of this, I think it'd be cleaner to just always initialize `update_graph` to an empty list, and then only on prod streams, we populate it. Then we can get rid of this check here since the for-loop below will automatically no-op.

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+        # Make sure this is not a barrier release (for stable streams)
+        if stream_id in [1, 2, 3]:
+            try:
+                for release in update_graph:
```
This is visiting all the releases on every iteration. Instead, we can build a list of all barrier releases once upfront. This should also allow us to get rid of this loop and the `BarrierRelease` exception trick.

jbtrystram replied:
I don't understand how I can get rid of the loop. I can build an array with all the barrier releases and do
```
if tag in barriers:
  continue
``` 

but that's still a loop (hidden admittedly)

jlebon replied:
You mean the loop to build the array? Yes indeed, but at least we'd be paying that cost once.

jbtrystram replied:
I did a think, let me know what you think !

jbtrystram replied:
> You mean the loop to build the array? Yes indeed, but at least we'd be paying that cost once.

I don't see how `if tag in barriers` would work without looping through the values :thinking:

jlebon replied:
> I don't see how if tag in barriers would work without looping through the values 🤔

Ahh gotcha, that hidden loop. Yes, I'm not too worried about the performance part, but more simplifying the code.

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,168 @@
+        if build_date < date_limit:
+            if args.dry_run:
+                print(f"Dry-run: would prune image {args.repository_url}:{tag}")
+                continue
```
Unnecessary

jlebon commented:
Also, typo in the commit and PR title: s/contaiers/containers/ :)

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,171 @@
+    parser.add_argument("--policy", required=True, type=str, help="Path to policy YAML file")
+    parser.add_argument("--dry-run", help="Don't actually delete anything", action='store_true')
+    parser.add_argument("-v", help="Increase verbosity", action='store_true')
+    parser.add_argument("--registry-auth-file", default=os.environ.get("REGISTRY_AUTHFILE"),
```
The canonical env var is `REGISTRY_AUTH_FILE` (see e.g. `podman-pull(1)`).

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,171 @@
+    # Load the policy file
+    with open(args.policy, "r") as f:
+        policy = yaml.safe_load(f)
+    if args.stream not in policy:
+        print(f"Stream {args.stream} is not defined in policy file.")
+        exit(1)
+    if 'containers' not in policy[args.stream]:
+        print(f"No containers section for {args.stream} stream in policy.")
+        exit(1)
```
```suggestion
    if args.stream not in policy:
        print(f"Stream {args.stream} is not defined in policy file; exiting...")
        return
    if 'containers' not in policy[args.stream]:
        print(f"No containers section for {args.stream} stream in policy; exiting...")
        return
```

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,171 @@
+    #    "Tags": [
+    #        "40.20"40.20240301.1.0",.....]}
+    tags_data = subprocess.check_output(["skopeo", "list-tags",
+                                        f"docker://{args.repository_url}"])
```
Minor/optional:

```suggestion
                                         f"docker://{args.repository_url}"])
```

Weird. I feel like flake8 used to warn about this but it doesn't anymore it seems like.

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,171 @@
+    if args.stream in ['stable', 'testing', 'next']:
+        update_graph = get_update_graph(args.stream)['releases']
+        # Keep only the barrier releases
+        # filter(lambda release: "barrier" in release["metadata"], update_graph)
```
Leftover?

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,171 @@
+    print(f"This will delete any images older than {date_limit} from the stream {args.stream}")
+
+    stream_id = STREAMS[args.stream]
+    barrier_releases = []
```
Minor/optional: hmm, I guess a set would be more appropriate here. Obviously totally inconsequential performance-wise for the scale we're at here, so feel free to leave as is.

jbtrystram replied:
Thanks for the suggestion, it's indeed a better fit

jlebon commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,171 @@
+        update_graph = get_update_graph(args.stream)['releases']
+        # Keep only the barrier releases
+        # filter(lambda release: "barrier" in release["metadata"], update_graph)
+        for release in update_graph:
+            if "barrier" in release["metadata"]:
+                barrier_releases.append(release["version"])
```
Minor/optional (related to above):

```python
barrier_releases = set([release["version"] for release in update_graph if "barrier" in release])
```

jbtrystram replied:
To me it's the same, but harder to read :D  But i am not well versed in python.
I applied the suggestion

jlebon replied:
That's fair. Wanted to show (what I think is) a more idiomatic way but also totally fine the more verbose way if it seems clearer to you.

jlebon reviewed (commented):
I think just the `exit(1)` bit needs fixing. The rest LGTM as is.

cverna commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,171 @@
+
+    unit = match.group(2)
+    value = int(match.group(1))
+    match unit:
```
you can probably do use unit.lower() to avoid having to match both lower and upper cases.

cverna commented on src/cmd-prune-containers:
```diff
@@ -0,0 +1,171 @@
+import requests
+import yaml
+
+# Dict of known streams
```
Does this have to be a dictionary? from a quit look a list might be enough.

jbtrystram replied:
yes because i need to associate the stream name with it's id. I tried to do a list of tuples but it was way more complicated.

jbtrystram commented:
Ok I rebased and updated this PR since https://github.com/coreos/coreos-assembler/pull/3798 landed. 

I made a couple of refactorings in `cmdlib.py` to share code between `cloud-prune` and this. 
@gursewak1997 PTAL at the duration parsing changes suggestion :)

jlebon commented on src/cmd-container-prune:
```diff
@@ -0,0 +1,125 @@
+    print(f"This will delete any images older than {date_limit} from the stream {args.stream}")
+
+    stream_id = STREAMS[args.stream]
+    barrier_releases = {}
```
Minor/optional:

```suggestion
    barrier_releases = set()
```

Feels weird to be casually changing types. :)
Then below, you can do `barrier_releases.update([release["version"] for ...])`.

jbtrystram replied:
I did `set()` below. 
Isn't `{}` a set as well ? I am confused

jlebon replied:
It's a dict:

```
$ python -c 'x = {}; print(type(x))'
<class 'dict'>
```

jbtrystram replied:
Ohh ok I probably misread the documentation

jbtrystram replied:
updated

jlebon reviewed (approved):
Nice! One comment, but LGTM as is!

gursewak1997 reviewed (approved):
Changes lgtm

gursewak1997 commented:
/test rhcos

jbtrystram commented:
/override ci/prow/rhcos

jlebon commented:
Prow failing on https://github.com/openshift/os/issues/1551.

/override ci/prow/rhcos
